mypy_django_plugin/transformers/managers.py

```diff
@@ -466,3 +466,55 @@ def create_new_manager_class_from_as_manager_method(ctx: DynamicClassDefContext)
         # Note that the generated manager type is always inserted at module level
         SymbolTableNode(GDEF, new_manager_info, plugin_generated=True),
     )
+
+
+def reparametrize_any_manager_hook(ctx: ClassDefContext) -> None:
+    """
+    Add explicit generics to manager classes that are defined without generic.
+
+    Eg.
+
+        class MyManager(models.Manager): ...
+
+    is interpreted as::
+        _T = TypeVar('_T', covariant=True)
+        class MyManager(models.Manager[_T]): ...
+    """
+
+    manager = ctx.api.lookup_fully_qualified_or_none(ctx.cls.fullname)
+    if manager is None or manager.node is None:
+        return
+    assert isinstance(manager.node, TypeInfo)
+
+    if manager.node.type_vars:
+        # We've already been here
+        return
+
+    parent_manager = next(
+        (base for base in manager.node.bases if base.type.has_base(fullnames.BASE_MANAGER_CLASS_FULLNAME)),
+        None,
+    )
+    if parent_manager is None:
+        return
+
+    is_missing_params = (
+        len(parent_manager.args) == 1
+        and isinstance(parent_manager.args[0], AnyType)
+        and parent_manager.args[0].type_of_any is not TypeOfAny.explicit
```

Comment:
There's also a `TypeOfAny.from_omitted_generics`. Not sure what we should use here, but we might perhaps be a bit more narrow if we do `TypeOfAny.from_omitted_generics`?

If it even works, might be cases I'm forgetting..

Reply:
Otherwise we could overwrite something like `TypeOfAny.from_error` etc etc

Reply:
Aight, I'll take a look and see what type of any we actually get in these cases and make it specific to that if possible :)

Reply:
I've updated this to only run if the any type is `TypeOfAny.from_omitted_generics`, which seems to work great 👍

Reply:
Correct me if I'm wrong, but I think what @sobolevn was trying to say is that even if we implicitly "repair" the omitted generic, mypy will still emit an error. Given you run with `--disallow-any-generics`

So one might argue that there's no need for us to fix these. Since one can find and make them explicit via mypy signaling where you've missed populating the generic by using the flag.

There's still a use case here though, and that's for our dynamically generated managers. In case they don't have any model arg populated. Since a user can't declare those explicitly

Reply:
Ah, okay. I'm following now, so basically we should remove this entire branch and accept that if you don't specify the generics of the class you'll get `Any`? I think one argument for keeping this third party libraries that are not typed. With this you'll at least get correct types on the queryset in your code.

I guess we could also put this behind a flag?

That said, I don't think we're really changing the behaviour of the mypy plugin here, it's always been doing this stuff, it's just been doing it using a different approach (creating a new class)

Reply:
It's important we do this for implicit managers from Django though (e.g. related managers)

Unless that's done someplace else

Reply:
Sorry for causing so much confusion :)

I was thinking about cases where our users have `disallow_any_generics = True` set, current implementation looks like it can still cause this error. Let's add a test case with `disallow_any_generics = True` set. I am pretty sure that we won't be able to fix / suppress this error, because it happens too early: https://github.com/python/mypy/blob/ddd917767aa8135f3b1aeef47b0bb0616a4b63fb/mypy/typeanal.py#L1462

But, we can at least be sure: what happens and what to expect.
Maybe even mention this in the docs 🤔

Reply:
Ah, okay :) I'll add a test case!

I don't think we should try to suppress the error message if possible though, if the user has enabled that flag I'd expect them to want a warning. At least in our codebase the implicit generics we add are less specific than what you'd manually add, so I consider this a sort of in-between solution until all of the project is fully typed (we're working in a fairly large codebase, gradually adding types)

Reply:
Hmm, it seems like this interacts a bit weirdly with `disallow_any_generics = True`. You'll get the expected warning from mypy and it seems like whatever we do in the hook is overwritten and the return types are changed back to `Any` 🤔  I added a test that demonstrates this behaviour

Reply:
Ah, I dug a bit deeper and when that option is enabled the `Any` type is changed to `TypeOfAny.from_error`, so the hook doesn't do anything. I think that's fair enough if you've enabled that option?